src/data/list/nodup.lean

```diff
@@ -127,6 +127,23 @@ theorem nodup_map_on {f : α → β} {l : list α} (H : ∀x∈l, ∀y∈l, f x
   (d : nodup l) : nodup (map f l) :=
 pairwise_map_of_pairwise _ (by exact λ a b ⟨ma, mb, n⟩ e, n (H a ma b mb e)) (pairwise.and_mem.1 d)
 
+theorem inj_on_of_nodup_map {f : α → β} {l : list α} (d : nodup (map f l)) :
+  ∀ ⦃x⦄, x ∈ l → ∀ ⦃y⦄, y ∈ l → f x = f y → x = y :=
+begin
+  induction l with hd tl ih,
+  { simp },
+  { simp only [map, nodup_cons, mem_map, not_exists, not_and, ←ne.def] at d,
+    rintro _ (rfl | h₁) _ (rfl | h₂) h₃,
+    { refl },
+    { apply (d.1 _ h₂ h₃.symm).elim },
+    { apply (d.1 _ h₁ h₃).elim },
+    { apply ih d.2 h₁ h₂ h₃ } }
+end
+
+theorem nodup_map_iff_inj_on {f : α → β} {l : list α} (d : nodup l) :
+  nodup (map f l) ↔ (∀ (x ∈ l) (y ∈ l), f x = f y → x = y) :=
```

Comment:
```suggestion
  nodup (map f l) ↔ (∀ ⦃x⦄, x ∈ l → ∀ ⦃y⦄, y ∈ l → f x = f y → x = y) :=
```

Reply:
To repeat my comment on zulip, I do not like the `{x : X} (h : x ∈ l)` pattern where only the proof term is explicit.  I find that this often makes it hard to write proofs incrementally.  For example `rw (nodup_map_iff_inj_on d).1 h x y` will rewrite `x` to `y` and give you two additional goals.  But `rw (nodup_map_iff_inj_on d).1 h` will just fail.  Using `@` to specify `x` and `y` is often inconvenient (when there are many type class arguments), or hard (like with the projection notation here).

I believe that `(x : X) (h : x ∈ l)` is always the better choice.